xds/src/main/java/io/grpc/xds/WrrLocalityLoadBalancer.java

```diff
@@ -0,0 +1,130 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+
+import com.google.common.base.MoreObjects;
+import io.grpc.InternalLogId;
+import io.grpc.LoadBalancer;
+import io.grpc.Status;
+import io.grpc.internal.ServiceConfigUtil.PolicySelection;
+import io.grpc.xds.WeightedTargetLoadBalancerProvider.WeightedPolicySelection;
+import io.grpc.xds.WeightedTargetLoadBalancerProvider.WeightedTargetConfig;
+import io.grpc.xds.XdsLogger.XdsLogLevel;
+import io.grpc.xds.XdsSubchannelPickers.ErrorPicker;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * This load balancer acts as a parent for the {@link WeightedTargetLoadBalancer} and configures
+ * it with a child policy in its configuration and locality weights it gets from an attribute in
+ * {@link io.grpc.LoadBalancer.ResolvedAddresses}.
+ */
+public class WrrLocalityLoadBalancer extends LoadBalancer {
+
+  private final XdsLogger logger;
+  private LoadBalancer childLb;
+  private final Helper helper;
+
+  WrrLocalityLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+    logger = XdsLogger.withLogId(
+        InternalLogId.allocate("xds-wrr-locality-lb", helper.getAuthority()));
+    logger.log(XdsLogLevel.INFO, "Created");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+
+    // The configuration with the child policy is combined with the locality weights
+    // to produce the weighted target LB config.
+    WrrLocalityConfig wrrLocalityConfig
+        = (WrrLocalityConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    Map<Locality, Integer> localityWeights = resolvedAddresses.getAttributes()
+        .get(InternalXdsAttributes.ATTR_LOCALITY_WEIGHTS);
+
+    // Weighted target LB expects a WeightedPolicySelection for each locality as it will create a
+    // child LB for each.
+    Map<String, WeightedPolicySelection> weightedPolicySelections = new HashMap<>();
+    for (Locality locality : localityWeights.keySet()) {
+      weightedPolicySelections.put(locality.toString(),
+          new WeightedPolicySelection(localityWeights.get(locality),
+              wrrLocalityConfig.childPolicy));
+    }
+
+    childLb = wrrLocalityConfig.childPolicy.getProvider().newLoadBalancer(helper);
```

Comment:
This creates a new childLb every update. We need to reuse the old LB. Options:
1) Assume the provider never changes after the first update
2) Check whether the new provider is equal to the old provider and recreate(+shutdown) the childLb if it changes. We commonly achieve this with the GracefulSwitchLB, which makes it easier.

Reply:
Thanks, I was thinking I am probably making some naive assumptions here. I updated this to use GracefulSwitchLoadBalancer, take a look.